### ISSUE TITLE
Unexpected error for generic intersection nested inside object

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#51555]
# Bug Report

### 🔎 Search Terms

- generic
- intersection
- objects

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about intersections

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UDesBcUDOwBOBLAdgOZQC+A3AFCiRQBCCyd6WeRplV40AwgPb4BjAIbAAPADkAjABoo4gEwA+BigD6AVQBKASXRSoAMjnz2FCgBMIAgDZDs0axGBQAHuj6CRomLNqLKljZ2Dk5QIO78wmK0sjD+ZgD0CVDq+BAukALAEOZQENjYvNioFElQAAqFkNigUADkMHVQuBhQALYtGATEBFDU0HW0TQBGAK7O9gCOo7j2ub399Y0AdBQuDCBkQA)

### 💻 Code

```ts
type A = { A: string };
type B = { B: string };

type Concat<N1, N2> = { _URI: N1 & N2 };

declare let x: Concat<A, B>;
declare let y: Concat<B, A>;

// Unexpected error:
// Property 'A' is missing in type 'B' but required in type 'A'.
x = y;
```

### 🙁 Actual behavior

See code comment above.

### 🙂 Expected behavior

See code comment above.

**Comment by RyanCavanaugh** [comment]
That's vexing as heck. `Concat` is measured to be covariant on its type parameters so we think we can do a variance-based measure, but that isn't fully complete as demonstrated here. I'm not sure what we could do without tanking performance - if anyone has a smart idea for a PR, I'd be very open to it, but I don't see a fix here.

Workaround:
```ts
type Flatten<T> = { [K in keyof T]: T[K] };
type Concat<N1, N2> = Flatten<{ _URI: N1 & N2 }>;
```
